synapse/http/client.py

```diff
@@ -21,73 +21,218 @@
     """
     A simple, no-frills HTTP client with methods that wrap up common ways of
     using HTTP in Matrix
     """
-    def __init__(self, hs):
+
+    def __init__(self, hs, treq_args={}, whitelist=None, blacklist=None):
+        """
+        Args:
+            hs (synapse.server.HomeServer)
+            treq_args (dict): Extra keyword arguments to be given to treq.request.
+            blacklist (netaddr.IPSet): The IP addresses that are blacklisted that
+                we may not request.
+            whitelist (netaddr.IPSet): The whitelisted IP addresses, that we can
+               request if it were otherwise caught in a blacklist.
+        """
         self.hs = hs
 
-        pool = HTTPConnectionPool(reactor)
-
-        # the pusher makes lots of concurrent SSL connections to sygnal, and
-        # tends to do so in batches, so we need to allow the pool to keep lots
-        # of idle connections around.
-        pool.maxPersistentPerHost = max((100 * CACHE_SIZE_FACTOR, 5))
-        pool.cachedConnectionTimeout = 2 * 60
-
-        # The default context factory in Twisted 14.0.0 (which we require) is
-        # BrowserLikePolicyForHTTPS which will do regular cert validation
-        # 'like a browser'
-        self.agent = Agent(
-            reactor,
-            connectTimeout=15,
-            contextFactory=hs.get_http_client_context_factory(),
-            pool=pool,
-        )
+        self._whitelist = whitelist
+        self._blacklist = blacklist
+        self._extra_treq_args = treq_args
+
         self.user_agent = hs.version_string
         self.clock = hs.get_clock()
+        self.reactor = hs.get_reactor()
         if hs.config.user_agent_suffix:
-            self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix,)
+            self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix)
 
         self.user_agent = self.user_agent.encode('ascii')
+        self._make_agent()
+
+    def _make_agent(self, _agent=False):
+
+        if _agent:
+            self.agent = _agent
+        else:
+            # the pusher makes lots of concurrent SSL connections to sygnal, and
+            # tends to do so in batches, so we need to allow the pool to keep
+            # lots of idle connections around.
+            pool = HTTPConnectionPool(self.reactor)
+            pool.maxPersistentPerHost = max((100 * CACHE_SIZE_FACTOR, 5))
+            pool.cachedConnectionTimeout = 2 * 60
+
+            # The default context factory in Twisted 14.0.0 (which we require) is
+            # BrowserLikePolicyForHTTPS which will do regular cert validation
+            # 'like a browser'
+            self.agent = Agent(
+                self.reactor,
+                connectTimeout=15,
+                contextFactory=self.hs.get_http_client_context_factory(),
+                pool=pool,
+            )
+
+        # If we have an IP blacklist, use the blacklisting Agent wrapper.
+        if self._blacklist:
```

Comment:
I find it a bit counterintuitive that the `_agent` param is ignored in this case. Suggestion: put whole the `if self._blacklist` block inside the `else` clause above

Reply:
done